### PR TITLE
[Core/docs] Add Simple AutoML with Core example

### PR DESCRIPTION
**Pull request by Yard1**
Signed-off-by: Antoni Baum <antoni.baum@protonmail.com>

## Why are these changes needed?

This PR adds a notebook showcasing how to create a simple AutoML system for time series using Ray Core. This is a common SSML use case we see with users. While this is possible using higher-level APIs (namely Tune), a self contained example like this is helpful for:

* Use cases not covered by our higher level APIs
* Use cases requiring higher performance than what we currently provide with higher-level APIs
* Showcasing the correct usage patterns of Ray Core primitives

Note: due to statsforecast conflicting with other requirements, this has to be ran in a separate buildkite job. Please let me know if you have ideas on how to avoid this (runtime envs, I guess? the entire notebook would need to be ran in one, but I just tested it and it seems to be working) and if this has to be avoided in the first place.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/ray-core/examples/automl_for_time_series.ipynb [hunk 1]
```diff
@@ -0,0 +1,592 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "@ray.remote\n",
+    "def train_and_evaluate_fold(\n",
```
Comment (jjyao) [inline doc/source/ray-core/examples/automl_for_time_series.ipynb:142]: Could we add more comments in the function saying what each line is doing? I think it's better to be verbose for examples.

File: doc/source/ray-core/examples/automl_for_time_series.ipynb [hunk 2]
```diff
@@ -0,0 +1,592 @@
...
+    "        return {metric_name: None for metric_name, metric in metrics.items()}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "`evaluate_models_with_cv` is a driver function to run our optimization loop. We take in a list of models (with their parameters already set) and the dataframe.\n",
+    "\n",
+    "The dataframe is put into the Ray object store and reused, which means we only need to serialize it once. That way, we avoid an {doc}`anti-pattern </ray-core/patterns/pass-large-arg-by-value>`.\n",
```
Comment (jjyao) [inline doc/source/ray-core/examples/automl_for_time_series.ipynb:172]: could we just do "{doc}`/ray-core/patterns/pass-large-arg-by-value`" so the title of the anti-pattern is included.

File: doc/source/ray-core/examples/automl_for_time_series.ipynb [hunk 3]
```diff
@@ -0,0 +1,592 @@
...
+    "In this notebook, we will build an AutoML (or more precisely, an AutoTS) system which will choose the best combination of a [statsforecast](https://github.com/Nixtla/statsforecast) model and hyperparameters for a time series regression task - here, we will be using a partition of the [M5 dataset](https://www.kaggle.com/c/m5-forecasting-accuracy)."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Simple AutoML consists of running different functions (hyperparameter configurations) on the same data independently of each other and choosing the best configuration according to a given metric. For a more practical example, we will be using [time series cross-validation (CV)](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.TimeSeriesSplit.html) as our evaluation strategy, which means that each hyperparameter configuration will be evaluated k-times, once for each fold of cross-validation.\n",
+    "\n",
+    "![AutoML](https://lh3.googleusercontent.com/JEKjXRUylChFqowbwts0Tdza0KIHnaJLKDiVmMjjRK_oCe9jUG3pGEq4ZPE9R3RrvB4=w2400)"
```
[the remark this answers is not in the document]
Comment (Yard1) [reply doc/source/ray-core/examples/automl_for_time_series.ipynb:31]: Yeah, I just created it :joy: let me check it in!

File: ci/env/install-dependencies.sh [hunk 4]
```diff
@@ -380,6 +380,13 @@ install_pip_packages() {
     pip install -U "ludwig[test]>=0.4" "jsonschema>=4"
   fi
 
+  # Additional dependency for statsforecast.
+  # This cannot be included in requirements_tune.txt as it has conflicting
+  # dependencies with Modin.
```
Comment (Yard1) [inline ci/env/install-dependencies.sh:385]:
```suggestion
  # dependencies.
```

**Comment by jjyao** [comment]
> Note: due to statsforecast conflicting with other requirements, this has to be ran in a separate buildkite job. Please let me know if you have ideas on how to avoid this (runtime envs, I guess? the entire notebook would need to be ran in one, but I just tested it and it seems to be working) and if this has to be avoided in the first place.

Seems runtime env is better so we can explicitly set the version of statsforecast that works for the example and doesn't need a whole new set of jobs/rules? @krfricke is there a better way?

btw, why do we need to that specific version (1.1.0) of statsforecast?

**Comment by Yard1** [comment]
@jjyao The statsforecast version is pinned following the practice we do with other requirements. It doesn't have to be specifically 1.1.0 to run, though.

My idea was to modify the script that runs the notebook as python code to run it in a Ray task with a runtime_env, but what happens is that the notebook starts a separate Ray instance. This works fine (the newly started Ray instance uses the runtime_env), but it seems quite hacky to me.

File: doc/source/ray-core/examples/automl_for_time_series.ipynb [hunk 5]
```diff
@@ -0,0 +1,616 @@
...
+    "        return {metric_name: None for metric_name, metric in metrics.items()}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "`evaluate_models_with_cv` is a driver function to run our optimization loop. We take in a list of models (with their parameters already set) and the dataframe.\n",
+    "\n",
+    "The dataframe is put into the Ray object store and reused, which means we only need to serialize it once. That way, we avoid an {doc}`/ray-core/patterns/pass-large-arg-by-value`.\n",
```
Comment (richardliaw) [inline doc/source/ray-core/examples/automl_for_time_series.ipynb:176]: Don't link to docs. Use internal refs.

File: doc/source/ray-core/examples/automl_for_time_series.ipynb [hunk 6]
```diff
@@ -0,0 +1,616 @@
...
+    "\n",
+    "We treat the fitting of each fold as a separate task. We generate k-tasks for each model, collect the metric results using {doc}`backpressure </ray-core/patterns/limit-tasks>` and aggregate them to calculate mean metrics.\n",
```
Comment (richardliaw) [inline doc/source/ray-core/examples/automl_for_time_series.ipynb:178]: same here.

File: doc/source/ray-core/examples/automl_for_time_series.ipynb [hunk 7]
```diff
@@ -0,0 +1,616 @@
...
+    "\n",
+    "In this notebook, we will build an AutoML (or more precisely, an AutoTS) system which will choose the best combination of a [statsforecast](https://github.com/Nixtla/statsforecast) model and hyperparameters for a time series regression task - here, we will be using a partition of the [M5 dataset](https://www.kaggle.com/c/m5-forecasting-accuracy)."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Simple AutoML consists of running different functions (hyperparameter configurations) on the same data independently of each other and choosing the best configuration according to a given metric. For a more practical example, we will be using [time series cross-validation (CV)](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.TimeSeriesSplit.html) as our evaluation strategy, which means that each hyperparameter configuration will be evaluated k-times, once for each fold of cross-validation.\n",
```
Comment (richardliaw) [inline doc/source/ray-core/examples/automl_for_time_series.ipynb:29]: After evaluation, we will be able to get XYZ?

File: doc/source/ray-core/examples/automl_for_time_series.ipynb (hunk 5 above)
Comment (jjyao) [reply doc/source/ray-core/examples/automl_for_time_series.ipynb:176]: Curious, why is that?

Comment (richardliaw) [reply doc/source/ray-core/examples/automl_for_time_series.ipynb]: If you move files around or move any sections away from the file, the links don't break. It's much more resilient to any slight refactoring.

**Review by richardliaw — APPROVED** [review]

File: doc/source/ray-core/examples/automl_for_time_series.ipynb (hunk 5 above)
Comment (jjyao) [reply doc/source/ray-core/examples/automl_for_time_series.ipynb]: I feel moving the file is similar to renaming the internal ref and is caught by our CI for any broken links. The fix is simple enough to be a search and replace.

I do like that when I link to a doc, sphinx will include the doc title...

**Comment by aslonnie** [comment]
the notebook is failing. the bucket `m5-benchmarks` is no longer accessible.

where does/did the `m5-benchmarks` live?

**Pull request merged**